### ISSUE TITLE
Easy to get m2e pom.xml editor fatally confused

### ISSUE THREAD

**Issue by davidmichaelkarr** [eclipse-m2e/m2e-core#1518]
I'm using Eclipse 2023-06.

In the last few days I've been seeing the m2e pom.xml editor get very confused, being unable to edit portions of the text, having it display blank lines where blank lines don't exist.  I find I have to close the editor and reopen the file. I have to be very careful entering text or it will get confused.

I am starting with this content:
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.att.idp</groupId>
  <artifactId>debugproblem</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  
</project>
```

For example, I can go to the blank line and enter TAB, and then enter the following:
```
<properties>>
```
By entering that errant ">", I've basically hosed that editor. If I try to erase one of those and then hit ENTER at the end of the line, it gets confused about what text is in the file and what it is displaying.  At this point I also see an error dialog like the following:
```
An error has occurred. See error log for more details.
startLine(5) does not match endLine(6)
```
The error log has this:
```
!ENTRY org.eclipse.ui 4 0 2023-08-22 14:40:10.449
!MESSAGE Unhandled event loop exception
!STACK 0
java.lang.IllegalStateException: startLine (5) does not match endLine (6)
	at org.eclipse.jface.text.projection.ProjectionMapping.toImageLine(ProjectionMapping.java:462)
	at org.eclipse.jface.text.TextViewer.modelLine2WidgetLine(TextViewer.java:5208)
	at org.eclipse.jface.text.source.MatchingCharacterPainter.handleDrawRequest(MatchingCharacterPainter.java:215)
	at org.eclipse.jface.text.source.MatchingCharacterPainter.paintControl(MatchingCharacterPainter.java:185)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:234)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4274)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1090)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1075)
	at org.eclipse.swt.widgets.Composite.WM_PAINT(Composite.java:1536)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:4800)
	at org.eclipse.swt.widgets.Canvas.windowProc(Canvas.java:340)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5040)
	at org.eclipse.swt.internal.win32.OS.DispatchMessage(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3658)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
```

**Comment by davidmichaelkarr** [comment]
It seems like it's even worse than this.  I'm starting from a valid pom.xml, and this is a snapshot of it:
![image](https://github.com/eclipse-m2e/m2e-core/assets/5566419/096b116c-d65e-4687-ae29-cb9316d32724)

I put my cursor at the end of the "`<exclusions>`" line and pressed Enter. Ok so far:
![image](https://github.com/eclipse-m2e/m2e-core/assets/5566419/0b0b0b19-af3e-452f-a4b6-74cc5182d93c)

I then entered "`<exclusion>`":
![image](https://github.com/eclipse-m2e/m2e-core/assets/5566419/02514bdd-2003-429f-9c2a-6201b7418c55)
Note that I did NOT press enter. After typing the ending "`>'", the text cursor stayed to the left of that character. If I try to move the cursor to the right, it jumps to the beginning of the next line. If I then try to move the cursor to the left, nothing happens.

I do note that if I accept the completion choices, it stays sane.  I usually take those recommendations, but sometimes I don't. I shouldn't expect that the editor will break if I don't.

**Comment by HannesWell** [comment]
@mickaelistria or @vrubezhny, as this is an Editor issue, could you please take a look?

**Comment by davidmichaelkarr** [comment]
By the way, I don't know if it matters, but if this is involving lsp, it might be useful to point out that I had filed an issue with SpringBoot, concerning some other issues with the Language Server: https://github.com/spring-projects/sts4/issues/1103 .

**Comment by HannesWell** [comment]
Thank you both for the detailed reports and testing.

> I did a lot of testing and I found a reproducible use case and also the **solution (or a workaround)** for it:
> **Switch back to `org.eclipse.lsp4e` version 0.16.1. The problem happens with 0.17.0., 0.17.1 and the latest 0.17.2**

In that case, can you please report this to the [lsp4e-project](https://github.com/eclipse/lsp4e)?
In general most of the Editor functionality often associated with m2e is actually provided by one of the following projects and therefore m2e cannot do much about bugs:
- https://github.com/eclipse/lemminx-maven
- https://github.com/eclipse/lsp4e
- https://github.com/eclipse-lsp4j/lsp4j
- https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper
- https://github.com/eclipse/tm4e

**Comment by chicobento** [comment]
Hi, was facing this error on eclipse 2023-06.
Just ran an update (Help->Check for updates) which updated m2e components / wildwebdeveloper and the bug is now gone.
Should this be closed ?

**Comment by HannesWell** [comment]
@davidmichaelkarr can you confirm this?

**Comment by davidmichaelkarr** [comment]
That will take a little while. I'm going to test this when I install 2023-09.

**Comment by chicobento** [comment]
Just an additional info: I ran another round of "Help->Check for updates" and now my 2023-06 installation has been upgraded to 2023-09 and the error is still gone. Thanks god, this was really annoying!

**Comment by mickaelistria** [comment]
Thanks for checking

**Comment by davidmichaelkarr** [comment]
Ok, I finally got 2023-09 set up and loaded up a project. I'm not seeing the error dialogs or the messed-up view. I can still get it moderately confused, although it's easy to recover from it.

I tried inserting a new "dependency" block.  At the proper place, I inserted exactly this text:
```
    <dependency>
```
As soon as I typed the closing ">", it produced this:
```
    <dependency></dependency></dependency>
```
I then manually removed the extra "</dependency>" to fix it. This was repeatable.
